### PR TITLE
Experimental OWASP automated threat and context threat generation

### PR DESCRIPTION
**Pull request by jgadsden**
**Summary**
This provides context based threat generation using the properties of the elements, such as 'Provides authentication' for an Actor element

**Description for the changelog**
experimental OWASP automated threat and context threat generation

**Other info**
Too big and rambling a change to be properly reviewed, so advised as 'experimental' in the application

### CODE REVIEW

**Comment by jgadsden** [comment]
@lreading , @andk123 , @mike-goodwin only review this is you feel like it. It is a huge feature spread over many files, but implements the experimental threats by context which use the properties of the elements.

These properties are now used for more than documenting the element

**Comment by andk123** [comment]
@jgadsden Will take some time this weekend to go over it 🙂. Need to refresh myself on the codebase as well!

File: td.desktop/core/services/joint.shapes.tm.js [hunk 1]
```diff
@@ -258,7 +258,7 @@ joint.shapes.tm.Process = joint.shapes.tm.toolElement.extend({
 
 //define process element properties
 
-joint.shapes.tm.utils.defineProperties(joint.shapes.tm.Process.prototype, ['privilegeLevel']);
+joint.shapes.tm.utils.defineProperties(joint.shapes.tm.Process.prototype, ['privilegeLevel', 'handlesCardPayment', 'isWebApplication', 'handlesGoodsOrServices']);
```
Comment (lreading) [inline td.desktop/core/services/joint.shapes.tm.js:261]: Are these all new properties being added to entities within the diagram? Should I include these and "storesInventory" in 2.0 as well?

**Comment by lreading** [comment]
Great stuff, Jon! I didn't pull/test the code, just took a look at the diffs and think I understand what's going on. :) Great addition!

**Review by lreading — APPROVED** [review]

File: td.desktop/core/services/joint.shapes.tm.js (hunk 1 above)
Comment (jgadsden) [reply td.desktop/core/services/joint.shapes.tm.js:261]: @lreading yes, good point, we needed to add some more context to the Process element and another to the Store element

If you could add these to version 2.0 also that would be great. I am vaguely worried that we might change our mind on some of these properties though, especially on the naming and the json entry

**Comment by jgadsden** [comment]
> @jgadsden Will take some time this weekend to go over it 🙂. Need to refresh myself on the codebase as well!

Thanks @andk123 , although it looks a bit scatter-gun it is all to do with the threat engine. This was an area @mike-goodwin suggested I look at in our video call way back in 2017 ... 4.5 years later finally got to it (apologies Mike)

**Comment by andk123** [comment]
LGTM, nice work!

However @jgadsden, quick question, when we add a threat based on the context, is this threat type always STRIDE by default (even if we select another type)? 

image

**Comment by jgadsden** [comment]
@andk123 that is a good point. The methodology and type are both TBD - I need to work out how to pick up on the diagram type and fix up the threat type ... will look at this next
Thanks!

**Comment by jgadsden** [comment]
@andk123 I have fixed up the model type and threat type for the context specific threats
Is it OK to merge now?

@lreading the threat engine will work same as before, with some fixup to the threat summary pane

@lreading and @andk123 I realise that this has added a large amount of translation to the localisation. We may want to stick with English for the threat  suggestions? Happy either way

**Review by lreading — APPROVED** [review]

**Comment by lreading** [comment]
> @andk123 I have fixed up the model type and threat type for the context specific threats Is it OK to merge now?
> 
> @lreading the threat engine will work same as before, with some fixup to the threat summary pane
> 
> @lreading and @andk123 I realise that this has added a large amount of translation to the localisation. We may want to stick with English for the threat suggestions? Happy either way

I've been going back and forth on translating the threats in v2, as I could see a desire for some to enter a custom threat type.  I think there may be an issue open for that, but cannot locate it at the moment.  That could change things, but not 100% sure the direction it'll take yet. :smile:

**Comment by andk123** [comment]
LGTM !

**Review by andk123 — APPROVED** [review]

**Pull request merged**